### DISCUSSION TITLE
Integration into @teambit

### DISCUSSION THREAD

**Discussion by Jinjiang** [vitest-dev/vitest#4772] [Q&A]
Hi, this is Jinjiang from @teambit. I'd like to create this discussion for the bit x vitest integration issues we have ever met.

_In short, Bit CLI is a tool for component development which chains linting, formatting, testing, and other tools together. In a Bit workspace, every component generates its own package in `node_modules`. For integrating vitest, most of the cases are just similar to normal frontend projects, except the watch mode since the content of packages those are generated by Bit components keep changing._

Here I list the issues we have met so far. Hopefully we can figure out a solution to make it works overall:

1. If the package is in ESM format, there are some behavior in Vitest different from in Vite:
    - The `.js` extension name in import statements has to be explicit.
    - The CSS/SASS/LESS files can't be imported
    - The static assets like images can't be imported
    - When there are both cjs and esm code in the package, Vitest will always find the cjs code rather than esm code.
      - Sometimes named export in the cjs code can't be found properly.

    Minimized reproduction and error logs: https://github.com/Jinjiang/reproductions/tree/vitest-20231218

2. We'd like to specify some packages in `node_modules` to be watched during the test.

3. In some cases, there are some prompts in terminal asking for installing dependencies like vitest ui, vitest coverage, jsdom etc. It would be great if we can catch them in Node.js APIs. So we can better deal with those case in the Bit workflow.

Thanks.

----

Some explanations: There was some discussions about named export in cjs before https://github.com/vitest-dev/vitest/issues/4745. And I have tried them all. However, I found:

- The `@apollo/client` is actually a esm package (with `type: module` in `package.json`). It just also includes cjs code, which by right I don't think should be accessed.
- `server.deps.inline` doesn't work with watch mode together I guess. I didn't find a way yet.

**Comment by sheremet-va** [comment]
> If the package is in ESM format, there are some behavior in Vitest different from in Vite:

Vitest externalizes packages that it thinks are correctly bundled. This means they are running with native Node.js ESM loader that doesn't support what you have. You need to use `server.deps.inline` to include those packages.

> When there are both cjs and esm code in the package, Vitest will always find the cjs code rather than esm code.

This is not true, Vitest uses a native Node.js ESM loader to find the code - if it's in `exports.import` then it will find it correctly. Vitest doesn't support "module" field since Node.js cannot load it.

All of the above is performance optimization that Vitest does. You can bypass it by using `server.deps.inline`.

> In some cases, there are some prompts in terminal asking for installing dependencies like vitest ui, vitest coverage, jsdom etc. It would be great if we can catch them in Node.js APIs.

How do you expect to catch them?

> * `server.deps.inline` doesn't work with watch mode together I guess. I didn't find a way yet.

This should actually work with watch mode, but you need to change watcher exclude. By default, it excludes all node_modules files.

**Comment by Jinjiang** [comment]
Thanks so much for all the information and advice. I updated the reproduction. Now the .js extension issue, CSS issue, and assets issue has been solved by `server.deps.inline`.

For watching files in `node_modules`, I tried multiple options but still didn't make it works. Would you mind taking a further look at https://github.com/Jinjiang/reproductions/blob/vitest-20231218/vite.config.js please? 🙏 

For `module` field in `package.json`, I understand that Node.js doesn't load according to it. And in the real case we are facing (`@apollo/client`) it doesn't behave differently. However, it theoretically runs to different code between in Vite and in Vitest. What do you think about that? 😅 

Thanks again.

**Reply by sheremet-va** [reply]
> For watching files in node_modules, I tried multiple options but still didn't make it works. Would you mind taking a further look at [Jinjiang/reproductions@vitest-20231218/vite.config.js](https://github.com/Jinjiang/reproductions/blob/vitest-20231218/vite.config.js) please?

I am not sure what you expect from `optimizeDeps`. Optimizer is disabled by default in Vitest. From what I can see watcher should _probably_ work, but I'm not sure. Vite always adds `**/node_modules/**` to exclude no matter what user provides.

> What do you think about that? 

This is needed to reliably run the tests, I don't know of any package that bundles completely different code for "main" and "module". We have already tried enabling it, but it breaks a lot of externalized packages which degrades performance. The current solution is the best possible middle ground. You can always pass down `module` flag using `resolve.mainFields` in your projects though.

**Reply by Jinjiang** [reply]
For the first issue, let me explain:

```js
{
server: {
  watch: {
    ignored: modules.map((m) => `!**/node_modules/${m}/**`)
  }
},
optimizeDeps: {
  exclude: modules
},
```

This is what I have tried so far the only way to watch the `modules` in `node_modules` in Vite. See the "Watching files in `node_modules`" tip on https://vitejs.dev/config/server-options.html#server-watch

```js
test: {
  watchExclude: modules.map((m) => `!**/node_modules/${m}/**`)
}
```

This is what I followed by the vitest docs https://vitest.dev/config/#watchexclude

I have tried each of them and all their combinations. They just didn't work in Vitest. Do you know how to make it works? 🙏

For the second issue, I will take the `resolve.mainFields` config necessarily. Thanks so much!

**Reply by sheremet-va** [reply]
> I have tried each of them and all their combinations. They just didn't work in Vitest. Do you know how to make it works? 🙏

I have found the issue, but I'm not sure I can fix it swiftly. Reproduction works correctly if you are using JSDOM or `happy-dom` environments thought. Please create a separate issue with the reproduction and mention it doesn't work with node environment only

**Reply by Jinjiang** [reply]
@sheremet-va the issue has been created https://github.com/vitest-dev/vitest/issues/4785
And I tested more about that. I found even in `jsdom` environment, the test won't be re-run in some cases. e.g. non-main file or the main file when you edit it the first time. I made another reproduction and describe more on the README and the issue I created. Thanks.

**Comment by Jinjiang** [comment]
> > In some cases, there are some prompts in terminal asking for installing dependencies like vitest ui, vitest coverage, jsdom etc. It would be great if we can catch them in Node.js APIs.
>
> How do you expect to catch them?

In Bit workspace, we don't use `npm/yarn/pnpm install` to install dependencies. Instead we have `bit install`. So it would be great if we can have an API, an option, or any kinds of event handlers which can be fired when any dependency is missing and let user decides whether to install it by Vitest or leave the job to user.

Pseudo code:

```js
startVitest(..., async function (dep, installFn) {
  if (needInstallByBit(dep)) {
    // manually install dep by Bit
  } else {
    await installFn(dep)
  }
}
```

Thanks.

**Reply by sheremet-va** [reply]
Please, open a separate feature proposal (via issues), so this thread is not forgotten.

**Reply by Jinjiang** [reply]
Sure. Here it is. Thanks 🙏 
https://github.com/vitest-dev/vitest/issues/4864
